Subject: zfcp: introduce BSG timeout callback
References: BNC#572659
From: Gerald Schaefer <geraldsc@de.ibm.com>
Patch-mainline: not yet

Symptom:     Kernel panic after a FC BSG requests is aborted while the
             corresponding zfcp request is still being processed.
Problem:     With zfcp, the underlying hardware cannot abort CT or ELS 
             requests, so there is nothing to do when the block layer 
             timeout expires.  
Solution:    To avoid interference with the block layer timeout, simply 
             indicate that the block layer timer should be reset. 
             The timer running in the hardware for the pending CT or ELS 
             request will return the request when it expires.
Acked-by: John Jolly <jjolly@suse.de>

---
 drivers/s390/scsi/zfcp_ext.h  |    1 +
 drivers/s390/scsi/zfcp_fc.c   |    6 ++++++
 drivers/s390/scsi/zfcp_scsi.c |    1 +
 3 files changed, 8 insertions(+)

--- a/drivers/s390/scsi/zfcp_ext.h
+++ b/drivers/s390/scsi/zfcp_ext.h
@@ -108,6 +108,7 @@ extern void zfcp_fc_wka_ports_force_offl
 extern int zfcp_fc_gs_setup(struct zfcp_adapter *);
 extern void zfcp_fc_gs_destroy(struct zfcp_adapter *);
 extern int zfcp_fc_exec_bsg_job(struct fc_bsg_job *);
+extern int zfcp_fc_timeout_bsg_job(struct fc_bsg_job *);
 
 /* zfcp_fsf.c */
 extern int zfcp_fsf_open_port(struct zfcp_erp_action *);
--- a/drivers/s390/scsi/zfcp_fc.c
+++ b/drivers/s390/scsi/zfcp_fc.c
@@ -767,6 +767,12 @@ int zfcp_fc_exec_bsg_job(struct fc_bsg_j
 	}
 }
 
+int zfcp_fc_timeout_bsg_job(struct fc_bsg_job *job)
+{
+	/* hardware tracks timeout, reset bsg timeout to not interfere */
+	return -EAGAIN;
+}
+
 int zfcp_fc_gs_setup(struct zfcp_adapter *adapter)
 {
 	struct zfcp_fc_wka_ports *wka_ports;
--- a/drivers/s390/scsi/zfcp_scsi.c
+++ b/drivers/s390/scsi/zfcp_scsi.c
@@ -652,6 +652,7 @@ struct fc_function_template zfcp_transpo
 	.show_host_port_state = 1,
 	.show_host_active_fc4s = 1,
 	.bsg_request = zfcp_fc_exec_bsg_job,
+	.bsg_timeout = zfcp_fc_timeout_bsg_job,
 	/* no functions registered for following dynamic attributes but
 	   directly set by LLDD */
 	.show_host_port_type = 1,
